Log location coordinates in the background

grzesiek2010 opened the pull request:
Closes #2773 
Closes #375 

#### What has been done to verify that this works as intended?
I wrote some tests and performed manual testing but it's not enough of course and the pr should be tested carefully by our QAs.

#### Why is this the best possible solution? Were any other approaches considered?
I tried to follow the approach agreed in https://github.com/opendatakit/roadmap/issues/28. Overall I was able to do everything as we discussed there apart from the section about `Data privacy issues` which I implemented in a bit different way.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
The pr brings a lot of changes in old time logging functionality which now allows logging location coordinates as well, so the biggest risk is related to that option.

The code that manages collecting events is in FormEntryActivity so there I also had to implement some changes but It doesn't need any specific regression testing.

#### Do we need any specific form for testing your changes? If so, please attach one.
[AuditTest.xml.txt](https://github.com/opendatakit/collect/files/2768473/AuditTest.xml.txt)

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
https://github.com/opendatakit/docs/issues/934

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

grzesiek2010 committed:
Created Audit class

grzesiek2010 committed:
Read audit parameters from forms and use that new Audit class instead of simple boolean

grzesiek2010 committed:
Factored out Namespaces class

grzesiek2010 committed:
Added tests for Audit class

grzesiek2010 committed:
Added a new field to check if a form is loaded for the first time

grzesiek2010 committed:
Pass just Audit object as a parameter instead of an entire FormController object

grzesiek2010 committed:
Use approprite column headers in audit file

grzesiek2010 committed:
Naming improvements - remove 'time' word since now logs are going to contain not only times but locations as well

grzesiek2010 committed:
Removed unused parameter from logEvent() method

grzesiek2010 committed:
Factored out Event class to make EventLogger class smaller

grzesiek2010 committed:
Edited Event class to handle location coordinates

grzesiek2010 committed:
Removed not needed default inizializations

grzesiek2010 committed:
Factored out comma char

grzesiek2010 committed:
Added tests for edited Event class

grzesiek2010 committed:
Collect locations during filling forms

grzesiek2010 committed:
Attach the most accurate location coordinates to an event

grzesiek2010 committed:
Removes collected locations that are too old

grzesiek2010 committed:
Use Long instead of Integer to store location-age

grzesiek2010 committed:
Added minimal acceptable location-interval

grzesiek2010 committed:
Added a dialog displayed after first load

grzesiek2010 committed:
Show snackbar if location coordinates should be displayed but it's not first load

grzesiek2010 committed:
Handle the case when Google Play Services are not available

grzesiek2010 committed:
Check phone state permission before calling the code related to collecting location coordinates

grzesiek2010 committed:
Moved the code related to collecting location coordinates to the bootom of the method which finishes loading forms

grzesiek2010 committed:
Got rid of the dialog displayed on first load

grzesiek2010 committed:
Check if location providers are enabled

grzesiek2010 committed:
Check location permission

grzesiek2010 committed:
Log an vent when location permissions are not granted

grzesiek2010 committed:
Added a button to snackbar to allow disabling gps

grzesiek2010 committed:
Check if Ggoogle Play Services are available before rquesting loaction permissions

grzesiek2010 committed:
Check if location permissions are granted as well in onStart() method

grzesiek2010 committed:
Removed redundant casts

grzesiek2010 committed:
Added license header

grzesiek2010 committed:
Fixed comment

grzesiek2010 committed:
Added license header

grzesiek2010 committed:
Added new icon to the manu - responsible for enabling/disabling collecting location coordinates

grzesiek2010 committed:
Added a new field to check if a form is loaded for the first time

grzesiek2010 committed:
Show the bacground location option only if Google Play Services are available

grzesiek2010 committed:
Improved data privacy messages

grzesiek2010 committed:
Improved messages to make them consistent

grzesiek2010 committed:
Setup location client only if background location is enabled

grzesiek2010 committed:
Start/stop location client after clicking on the new option in the menu

grzesiek2010 committed:
Perform the same steps after enabling bacground location as in case of form load

grzesiek2010 committed:
Added LOCATION_PROVIDERS_DISABLED event

ggalmazor commented:
I've tried this PR on my android device and I get some extra columns on my audit file with the latitude, longitude and accuracy.

What position does it correspond to? The first, the last, or another position?

I also see that there is no way to manually dismiss the message. Just saying that in case it was not the pretended behavior.

Also: Aggregate may need a little change in order to show the new columns on the audit popup. Shall we create an issue for that?

grzesiek2010 committed:
Added missing implementation of BACKGROUND_LOCATION_ENABLED and BACKGROUND_LOCATION_DISABLED events

grzesiek2010 commented:
@ggalmazor 
the pr is not ready yet. Sorry, I need to and `WIP` back. Last Friday we agree on some last details: https://github.com/opendatakit/roadmap/issues/28 I'll update the pr today or tomorrow once I finish.

grzesiek2010 committed:
Removed unused field

grzesiek2010 committed:
Moved event log up

grzesiek2010 committed:
Log an event if a user grants/revokes location permissions during form filling

grzesiek2010 committed:
Log an event if a user enables/disables location providers during form filling

grzesiek2010 committed:
Fixed bug in GoogleLocationClient - setInterval called two times instead of setFastestInterval

grzesiek2010 committed:
Init background location after starting saved form

grzesiek2010 committed:
Fixed bug in removeExpiredLocations() method

grzesiek2010 committed:
Update header in audit file if new form version requires location coordinates

grzesiek2010 commented:
### What is the general goal of the feature?
As a supervisor, I would like to collect location coordinates in the background to provide evidence that data was collected in a particular place.
The reason of this feature is to avoid bad data. For example, you might notice an enumerator who as been poorly trained collect location in the morning, then go home and fill in the survey data whenever.
Implementation

### Implementation
The implementation extends [form audit log](https://docs.opendatakit.org/form-audit-log/) adding location to the log if all three:

- **location-priority**: no-power, low-power, balanced, high-accuracy as defined in [LocationRequest](https://developers.google.com/android/reference/com/google/android/gms/location/LocationRequest)
- **location-interval**: the desired minimum interval in seconds that the location will be fetched
- **location-age**: the time in seconds that will location will be valid

are set as parameters in the audit row of the XLSForm.
Currently an audit.csv file has the following structure:

event | node | start | end
-- | -- | -- | --
question | /data/name | 1523403169208 | 1523403170733

Location will be recorded in three columns: latitude, longitude, accuracy in the audit log that is attached to Collect's submissions.
New structure of an audit.csv file:


event | node | start | end | latitude | longitude | accuracy
-- | -- | -- | -- | -- | -- | --
question | /data/name | 1523403169208 | 1523403169208 | 27.1080 | 11.19574 | 20

- New columns will be added only if the audit row of the XLSForm contains background parameters otherwise they are not needed.
- When you have an existing form with audit that did not have location enabled and you change that form to add location, three new columns will be added.

### XForms specification
In the XForms spec audit is placed in the meta block under the orx namespace. It uses a binary data type. See [opendatakit/xforms-spec#94](https://github.com/opendatakit/xforms-spec/issues/94)
```xml
<orx:meta>
	<orx:audit/>
</orx:meta>
<bind nodeset="/orx:meta/orx:audit" type="binary"/>
```
We will add the following bind attributes in the odk namespace to the specification. Both attributes are required to enable location audits and location-age must be greater than or equal to location-interval.
```xml
<bind nodeset="/orx:meta/orx:audit" type="binary" 
           odk:location-priority="balanced" odk:location-interval="10" odk:location-age="60" />
```
**Sample form:** 
[AuditTest.xml.txt](https://github.com/opendatakit/collect/files/2768475/AuditTest.xml.txt)

### Device behavior
When Collect opens the form, we will request location updates from the location provider every desired interval seconds. Location updates will be started in once the form is loaded (loadingComplete() method) and paused when the app is in the background (onStop() method/started again in onStart()) to avoid battery draining.
At the time of the [audit event](https://docs.opendatakit.org/form-audit-log/#event-types) (e.g., view question, form save etc.) occurs, Collect will check the cache for the location and add coordiantes to the event.

**Sample:**
In the ideal case for location-interval=10 and location-age=60, 6 points will be collected and we choose the most accurate one to write the log. But location-interval is not guaranteed and so there may be more points or less points.

In order to allow our users using such a form without collecting location coordinates we agreed on adding a new option to the menu:
![screenshot from 2019-01-09 12-53-41](https://user-images.githubusercontent.com/3276264/50897963-fa6ba600-140d-11e9-889b-037d865fabea.png)
![screenshot from 2019-01-09 12-54-00](https://user-images.githubusercontent.com/3276264/50897964-fb043c80-140d-11e9-8f24-e353c8fbbd14.png)

of cours it's also possible to block colecting location coordinates by revoking location permissions on Android 6+ or disabling location providers but it's not a convinient way.

### Data privacy
To ensure enumerators are aware that this sensitive data is being gathered:

- A user opens a form which needs to collect location in the background.
1. Check if Google Play Services are available, if not display a Snackbar:
`This form needs to collect your location in the background, but it's not possible because Google
Play Services are not available.`
and log an event: `GOOGLE_PLAY_SERVICES_NOT_AVAILABLE`
otherwise, go to the next step...
2. Check if the background location option is enabled, if not display a Snackbar:
`This form needs to collect your location in the background, but the background location option is disabled. You may enable this feature in the menu.`
and log an event: `BACKGROUND_LOCATION_DISABLED`
otherwise, log an event: `BACKGROUND_LOCATION_ENABLED` and go to the next step...
3. Request location permissions, if not granted log an event:
`LOCATION_PERMISSIONS_NOT_GRANTED`
otherwise, log an event:
`LOCATION_PERMISSIONS_GRANTED` and go to the next step...
4. Check if location providers are enabled, if not display a Dialog:
`This form needs to collect your location in the background, but location providers are disabled. You may enable this feature in the settings.`
with an option `Go to settings`.
and log an event `LOCATION_PROVIDERS_DISABLED`
otherwise, show a Snackbar:
`This form collects your location continuously in the background. You may disable this feature in the menu.`
and log location coordinates!
 
- A form is started and a user disables background location 
stop the location client and log an event `BACKGROUND_LOCATION_DISABLED`

- A form is started and a user enabes background location 
log an event `BACKGROUND_LOCATION_ENABLED`
every step like in the Case 1 (from step 3)

- A form is started and a user revokes location permissions 
log an event `LOCATION_PERMISSIONS_NOT_GRANTED`

- A form is started and a user grants location permissions again
log an event `LOCATION_PERMISSIONS_GRANTED`

- A form is started and a user disables location providers 
log an event `LOCATION_PROVIDERS_DISABLED`

- A form is started and a user enables location providers back
log an event `LOCATION_PROVIDERS_ENABLED`

grzesiek2010 commented:
@yanokwa @ggalmazor @shobhitagarwal1612 @zestyping and everyone who is interested in reviewing and testing this pr: I've just added a more detailed description for you to help you understand everything.
It's a sum up because reading our long discussions
https://github.com/opendatakit/roadmap/issues/28 might be boring and misleading.

grzesiek2010 committed:
Record enabling/disabling location providers using Broadcastreceiver

mmarciniak90 commented:
Because it is a quite big feature, we started testing before final review.

Noticed:
1. Information in END column are lost probably when the user makes some changes with location
<img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/51031662-8b24bc00-159d-11e9-83a2-9837b51a76a2.png">

2. Location is check when the form is loaded because then we know that we should check location in the background or not. Location is set in Audit file when some event occurs.
That cause first question in form does not have location information.

<img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/51031662-8b24bc00-159d-11e9-83a2-9837b51a76a2.png">

* If all questions are visible on **one page**, the same case is visible
[AuditTestOnePage.xml.txt](https://github.com/opendatakit/collect/files/2749444/AuditTestOnePage.xml.txt)

<img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/51031692-a263a980-159d-11e9-87a5-52461d4f5339.png">

kkrawczyk123 commented:
I think that it would be great if those extra locations columns in audit file will be also displayed on aggregate as now they are not and it currently looks like:
![screenshot from 2019-01-11 15-46-13](https://user-images.githubusercontent.com/33342488/51040630-4f4b2000-15b8-11e9-93c3-51110e620dcd.png)
maybe @ggalmazor could help with that somehow.

grzesiek2010 committed:
Fixed empty end time when location events happen

grzesiek2010 commented:
> 1. Information in END column are lost probably when the user makes some changes with location

Fixed.

> 2. Location is check when the form is loaded because then we know that we should check location in the background or not. Location is set in Audit file when some event occurs.
That cause first question in form does not have location information.

I could fix it by adding locations when the `end time` is set (once we leave a question). Then in most cases, it would be enough time to collect coordinates but I'm not sure we didn't talk about it and I assumed that for all events, location should be added at the same time the event is logged no matter if it's an internal event like question where we keep such an event and write it to the audit file once it's finished.

@yanokwa what do you think?

> 3. If all questions are visible on one page, the same case is visible

Yes if it's one page we don't log events for each question. I don't think it's a big problem since I doubt someone uses big forms (where logging locations might be needed) on one page. So I would leave it as it is for now.

yanokwa commented:
@grzesiek2010 I agree with you that `form start` can't really have a location. We have to load the form first before we ask all the location questions. For multiple questions per screen, the behavior should match the time logging. If this is hard to do or dramatically makes the code more complex, then we should discuss it.

@kkrawczyk123 I hear you that Aggregate doesn't display the data, but does it download correctly?

@mmarciniak90 Forgive me if this written somewhere, but it would be good to know what specific scenarios you've verified so far.

yanokwa commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -1152,10 +1154,17 @@ public InstanceMetadata getSubmissionMetadata() {
-                audit = true;
+                TreeElement auditElement = v.get(0);
+
+                String locationPriority = auditElement.getBindAttributeValue(XML_OPENDATAKIT_NAMESPACE, "location-priority");
```
These should probably end up  in JavaRosa, no?

yanokwa commented on collect_app/src/main/java/org/odk/collect/android/logic/Audit.java:
```diff
@@ -0,0 +1,88 @@
+            case "no_power":
+            case "no-power":
+                return LocationClient.Priority.PRIORITY_NO_POWER;
+            default:
```
I'd prefer that balanced be the fall through and high-accuracy be the default.

kkrawczyk123 commented:
@yanokwa, yes after pulling it looks fine.

mmarciniak90 commented:
@yanokwa List of verified scenarios will be added during the final testing. 
When I do exploratory testing on a new feature I'm not focused on creating test cases but more on finding defects. Especially, because testing is not prepared on all Android version but only on few selected.

grzesiek2010 committed:
Add location when end time is set if it's an interval event and the location is not already set

smoyte commented:
Sorry I'm a bit late to the party here. I was asked at the last TSC meeting to review the UX here. My comments are from the enumerator's perspective, not the user consuming the audit data (I looked briefly at the CSV file and it looks fine but I'm not looking at how it shows up in Aggregate.)

I really appreciate all the thought and work that went into this feature!

The flow from system permission request > snackbar is nice.

The message in the snackbar is "This form collects your location continuously in the background. You may disable this feature in the menu." This was confusing at first. I had to figure out that the map marker icon is what controls this, and that icon is not really "in" a "menu", it's in the action bar.

But before we get into this, I am not in favor of using up valuable action bar real estate with this toggle icon. It should be used rarely. I also worry about it interfering with muscle memory of where the jump screen icon is. Can we not just do as the snackbar says and put it in the popup menu? This would make it more discoverable as well, not requiring you know the map location icon meaning.

It could be a checkable popup menu item called "Location Tracking". Something like this:

![image](https://user-images.githubusercontent.com/702320/51192688-1dd29d00-18b5-11e9-877c-b66574339321.png)

If you disable and re-enable location tracking, I don't think we need to show the snackbar again. By then the user has figured out how it works, etc. Can we just show these on form load?

The snackbars disappear kind of quickly sometimes. Especially the longer one. Can we increase the timeout a bit?

Sorry if I have repeated any earlier comments. I just don't have time to read the massive back and forth from earlier. On the plus side, I'm coming at it with fresh eyes :)

Really great work here!

grzesiek2010 committed:
Show snackbars with messages about background gps longer

grzesiek2010 committed:
Don't show a snackbar when background location option is enabled again by a user after loading a form

grzesiek2010 commented:
@mmarciniak90 @yanokwa 
I solved the problem with empty location coordinates in the first question.
Location coordinates are added to each event once that event happens but then in case of questions, we set end time once we exit a question, at that time I check if location coordinates were set and if not I add them.
It's not a perfect solution because if someone swipes quickly it might be still empty but it improves it a little bit.

@smoyth 
>If you disable and re-enable location tracking, I don't think we need to show the snackbar again. By then the user has figured out how it works, etc. Can we just show these on form load?

I agree, fixed.

>The snackbars disappear kind of quickly sometimes. Especially the longer one. Can we increase the timeout a bit?

You are right I improved it. It was 3.5s and now it's 7s

When it comes to the option to enabling/disabling background location if think we have to discuss it.
@yanokwa what do you think?

zestyping commented on collect_app/src/test/java/org/odk/collect/android/logic/AuditTest.java:
```diff
@@ -0,0 +1,100 @@
+        Audit audit = new Audit("high-accuracy", "10", "60");
+
+        assertTrue(audit.collectLocationCoordinates());
+        Assert.assertEquals(PRIORITY_HIGH_ACCURACY, audit.getLocationPriority());
```
`assertEquals` is used so often that it would be nice to have a static import for it, like the one in `EventTest`.

zestyping commented on collect_app/src/test/java/org/odk/collect/android/logic/EventTest.java:
```diff
@@ -0,0 +1,171 @@
+import static org.odk.collect.android.utilities.EventLogger.EventTypes.BACKGROUND_LOCATION_ENABLED;
+import static org.odk.collect.android.utilities.EventLogger.EventTypes.CONSTRAINT_ERROR;
+import static org.odk.collect.android.utilities.EventLogger.EventTypes.DELETE_REPEAT;
+import static org.odk.collect.android.utilities.EventLogger.EventTypes.FEC;
```
`FEC` is a mysterious and cryptic name for a constant.  I wish it were named something like `FORM_EVENT` or `FORM_NAVIGATION`.  Perhaps that's out of scope for this PR though.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/logic/Audit.java:
```diff
@@ -0,0 +1,88 @@
+import io.reactivex.annotations.NonNull;
+import io.reactivex.annotations.Nullable;
+
+public class Audit {
```
This class needs an overview comment to explain what an `Audit` is.

A more specific name would also be good.  The name `Audit` isn't enough for the reader to tell what role this class plays in an audit, e.g. whether it's the output of an audit, or a thing that performs an audit, or a thing that generates an audit log, etc. etc.  This looks like an audit configuration, if so, perhaps `AuditConfig`?

yanokwa commented:
@grzesiek2010 I think I'd misread the initial complaint from @mmarciniak90. It's not the first event that has no location, it's the first question. In that case, I think your solution is a good one, but I think some consistency there would be nice. That is, if we have an end time, we write the location then. So you'd have to add support for group question, add/delete repeat, end screen. This also makes the documentation easier to write.

As to the disable/enable UI, I played with it a bit more today and I think moving the option to the action bar would be better. Maybe above Change Language? I don't how hard it'll be to add a toggle, but perhaps you can look into it and see.

Finally, I tried the current changes with @lognaturel and she thought the messages are too long, even with the 7 second snackbar. Maybe we can shorten them? What do you think of these changes, @smoyth? A friendlier alternative would be that we show a blocking dialog the first time, then show a snackbar after that.

**original**
```xml
<string name="google_play_services_not_available">This form needs to collect your location in the background, but it\'s not possible because Google Play Services are not available.</string>
<string name="location_providers_disabled_dialog_message">This form needs to collect your location in the background, but location providers are disabled. You may enable this feature in the settings.</string>
<string name="background_location_disabled">This form needs to collect your location in the background, but the background location option is disabled. You may enable this feature in the menu.</string>
<string name="background_location_enabled">This form collects your location continuously in the background. You may disable this feature in the menu.</string>
```

**proposed**
```xml
<string name="google_play_services_not_available">Background location required, but Google Play Services are not available.</string>
<string name="location_providers_disabled_dialog_message">Background location required, but location providers are disabled. Enable in Android settings.</string>
<string name="background_location_disabled">Background location required, but is disabled. You may enable in ⋮ menu at top of screen.</string>
<string name="background_location_enabled">Background location enabled. You may disable in ⋮ menu at top of screen.</string>
```

zestyping commented on collect_app/src/main/java/org/odk/collect/android/logic/Event.java:
```diff
@@ -0,0 +1,167 @@
+import org.javarosa.form.api.FormEntryController;
+import org.odk.collect.android.utilities.EventLogger;
+
+public class Event {
```
`Event` is a pretty generic name.  This class deserves an overview comment to tell us a little about what these events are for.  You might also want to consider a name that helps to distinguish it from the many other kinds of events.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/logic/Event.java:
```diff
@@ -0,0 +1,167 @@
+
+        String log = textValue + COMMA + node + COMMA + start + COMMA + (end != 0 ? end : "");
+        if (collectLocationCoordinates) {
+            log += COMMA + latitude + COMMA + longitude + COMMA + accuracy;
```
Consider using `String.format` on lines 161 and 163?

zestyping commented on collect_app/src/main/java/org/odk/collect/android/logic/Event.java:
```diff
@@ -0,0 +1,167 @@
+import org.odk.collect.android.utilities.EventLogger;
+
+public class Event {
+    private static final String COMMA = ",";
```
This constant seems unnecessary to me.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/logic/Event.java:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    /*
+     * Return true if this is a view type event
```
I'm having trouble understanding what a "view type event" is, or what the name "isIntervalViewEvent" means.  Does this just mean "an event that has both a start time and end time"?  If so, perhaps `isIntervalEventType` would be clearer.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/logic/Event.java:
```diff
@@ -0,0 +1,167 @@
+                        textValue = "Unknown FEC: " + fecType;
+                }
+                break;
+            case FORM_START:
```
That's a lot of cases.  Would you prefer to use an EnumMap<EventType, String> instead, so this becomes `eventType = eventNames.getOrDefault(eventType, "Unknown...")` ?

zestyping commented on collect_app/src/main/java/org/odk/collect/android/logic/Event.java:
```diff
@@ -0,0 +1,167 @@
+    private String latitude;
+    private String longitude;
+    private String accuracy;
+    private boolean collectLocationCoordinates;
```
I'm wondering what the intended semantics of `collectLocationCoordinates` are.

Is it meant to indicate whether location information is present?  If so, `hasLocation` seems like a better name (there's no collecting happening here).

Does its value always match `areLocationCoordinatesSet()`?  If so, I'd rename that method to `hasLocation()` and delete this field.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/utilities/EventLogger.java:
```diff
@@ -1,31 +1,40 @@
+ * 2) The events for questions in a field-list group are not shown.  Only the event for the group is shown.
  */
-public class TimerLogger {
+public class EventLogger {
```
It feels to me like `EventTypes` would belong better inside `Event` (as `Event.Type`) than inside `EventLogger`.  What do you think?

zestyping commented on collect_app/src/main/java/org/odk/collect/android/utilities/EventLogger.java:
```diff
@@ -259,17 +167,39 @@ public void logTimerEvent(EventTypes eventType,
+
+    // If location provider are enabled/disabled it sometimes fires the BroadcastReceiver multiple
+    // times what tries to add duplicated logs
+    private boolean duplicatedLocationProvidersLog(EventLogger.EventTypes eventType) {
```
This method name is a bit confusing to me, partly because it doesn't sound like a condition and partly because "log" means the log of all the events, whereas this is about a single event.  I think you mean to check whether the given event type is a duplicate of the last recorded event, is that right?  If that's right, how about a name like `isDuplicateOfLastEvent()`?

zestyping commented on collect_app/src/main/java/org/odk/collect/android/utilities/EventLogger.java:
```diff
@@ -305,4 +235,38 @@ private long getEventTime() {
+                } else {
+                    break;
+                }
+            }
```
This could be simplified by keeping only the locations you want to keep, instead of copying them all and then removing some.

```
List<Location> unexpiredLocations = new ArrayList<>();
long now = System.currentTimeMillis();
long maxAge = audit.getLocationAge();
for (Location location : locations) {
    if (now < location.getTime() + maxAge) {
        unexpiredLocations.add(location);
    }
}
locations = unexpiredLocations;
```

zestyping commented on collect_app/src/main/java/org/odk/collect/android/utilities/EventLogger.java:
```diff
@@ -305,4 +235,38 @@ private long getEventTime() {
+                    bestLocation = location;
+                }
+            }
+        }
```
If you like lambdas, this could be simplified to

```
if (!locations.isEmpty()) {
    return Collections.min(locations, Comparator.comparing(location -> location.getAccuracy()));
```

grzesiek2010 replied:
Would be nice but Comparator.comparing requires API 24

zestyping commented on collect_app/src/main/java/org/odk/collect/android/logic/Audit.java:
```diff
@@ -0,0 +1,88 @@
+        return locationAge;
+    }
+
+    public boolean collectLocationCoordinates() {
```
Because this method name starts with the verb "collect", it sounds like an action (as though calling this method will collect something).  For a Boolean check like this, it's better to make it sound like a condition, e.g. `isLocationEnabled`.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/logic/Audit.java:
```diff
@@ -0,0 +1,88 @@
+    /**
+     * The time in milliseconds that location will be valid
+     */
+    private final Long locationAge;
```
This represents not the age of a particular location reading, but the maximum allowed age for any location reading.  So it would be better to name this field `locationMaxAge` instead of `locationAge`.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -1152,10 +1154,17 @@ public InstanceMetadata getSubmissionMetadata() {
+
+                String locationPriority = auditElement.getBindAttributeValue(XML_OPENDATAKIT_NAMESPACE, "location-priority");
+                String locationInterval = auditElement.getBindAttributeValue(XML_OPENDATAKIT_NAMESPACE, "location-interval");
+                String locationAge = auditElement.getBindAttributeValue(XML_OPENDATAKIT_NAMESPACE, "location-age");
```
I seem to recall that we had discussed this at the spec stage and decided to name the XML attribute "location-max-age" instead of "location-age".  @yanokwa Can you confirm?

yanokwa replied:
Good catch. I confirm that we'd agreed on `min-interval` and `max-age`

grzesiek2010 replied:
Ok so I'm confused...
`location-priority, location-min-interval, location-max-age`
or
`priority, min-interval, max-age `
?

yanokwa replied:
Sorry I wasn't clear! `location-priority`, `location-min-interval`, `location-max-age`

zestyping commented on collect_app/src/main/java/org/odk/collect/android/tasks/EventSaveTask.java:
```diff
@@ -0,0 +1,113 @@
+ */
+public class EventSaveTask extends AsyncTask<Event, Void, Void> {
+    private final @NonNull File file;
+    private final boolean collectLocationCoordinates;
```
Similar to a comment I made elsewhere—there's no collecting happening in this class, and naming this boolean to sound like a condition would help; how about `isLocationEnabled` or `locationOutputEnabled`?

zestyping reviewed (commented):
Hi, I had a look through this and have a few comments.  Many are about naming, most are quick changes if you agree.  Hope this is helpful!

Regarding testing:
- It would be great to have a unit test for `EventLogger`—is that possible?
- `EventSaveTask` would be good to test as well, though it's harder to do that—you would need to factor out pieces of it into testable methods.

grzesiek2010 committed:
Removed unused method

grzesiek2010 committed:
Added static Assert.assertEquals import

grzesiek2010 committed:
Initialize events array on declaration

grzesiek2010 committed:
Removed nested if statemenst

grzesiek2010 committed:
Removed redundant constant

grzesiek2010 committed:
Removed redundant collectLocationCoordinates field

grzesiek2010 committed:
Removed redundant null check

grzesiek2010 committed:
Use fecType = 0 by default parameter if not needed

grzesiek2010 committed:
Use String.format to create logs

grzesiek2010 committed:
Method name improvement

grzesiek2010 committed:
Field name improvement

grzesiek2010 committed:
Added NonNull annotation

grzesiek2010 committed:
Improved removeExpiredLocations() method

grzesiek2010 committed:
Methoda neme improvement

grzesiek2010 committed:
Moved EventTypes to Event class

grzesiek2010 committed:
Removed FEC event and fecType

smoyte commented:
@yanokwa It sounds like we agree on moving the toggle into a popup menu in the action bar. But just to clarify, the icon toggle _is already in the action bar_. I'm just suggesting moving it into the popup menu. Right???

Regarding strings, @yanokwa I think your changes lose some important bits of meaning. How about:

```
<string name="google_play_services_not_available">This form wants to track your location but Google Play Services are not available.</string>
<string name="location_providers_disabled_dialog_message">This form wants to track your location but location providers are disabled. Please enable in Android settings.</string>
<string name="background_location_disabled">This form wants to track your location but tracking is disabled. Please enable in the ⋮ menu above.</string>
<string name="background_location_enabled">This form tracks your location. You can disable tracking in the ⋮ menu above.</string>
```

I think it's important to relate the tracking request to the form, not _all_ forms. This is not some new feature of Collect that tracks location on all forms.

I also think we need to be subtle with language about need vs. want. The form 'wants' to track but doesn't need to. It will still work without. We say 'please' enable as a firm request/recommendation, knowing it might technically be ignored.

grzesiek2010 committed:
Improved reading event names

grzesiek2010 committed:
Fixed tests

grzesiek2010 committed:
Added unknow event type

grzesiek2010 committed:
Improved toString method in Event class

grzesiek2010 committed:
Removed empty lines

grzesiek2010 committed:
Removed redundant auditEnabled field

grzesiek2010 committed:
Improved parameter name and added a comment

grzesiek2010 committed:
Fixed audit parameter names

grzesiek2010 committed:
Renamed Audit class to make it more meaningful

grzesiek2010 committed:
Renamed Event classes to make names more meaningful

grzesiek2010 committed:
Renamed method

grzesiek2010 committed:
Use an empty array to convert a collection to an array

grzesiek2010 committed:
Fixed naming bugs

grzesiek2010 commented:
@zestyping Thanks for your comments.
Please review new commits, I think I fixed everything.
You also asked me for adding more tests and I'll try to do that once everything else is fine and accepted.

zestyping commented:
@grzesiek2010 Thanks so much, this all looks great!  I really like your solution to the FEC event type (conversion in getEventType), and you put the event type name in the enum, which is way better than the HashMap idea I suggested.

zestyping commented:
@grzesiek2010 Oh, I noticed one thing, though.  Java type names should be singular; the enum should be `EventType` rather than `EventTypes`.

It makes sense if you imagine a method declaration:
```
    void foo(EventTypes x)  // but foo takes one event type, not multiple
    void foo(EventType x)   // better
```

yanokwa commented:
@smoyth I should be more precise. Yes, it's already in the action bar. We just want to move it into the pop up menu. @grzesiek2010, have you been able to explore if that is doable?

@grzesiek2010, can you please update the text to what @smoyth suggested at https://github.com/opendatakit/collect/pull/2772#issuecomment-454824524.

We are very close to finishing this feature up. Exciting!

zestyping commented:
I was going through geo-related issues and noticed that #375 looks related.  Does this close that issue also?

grzesiek2010 committed:
Type name should be singular

grzesiek2010 committed:
Updated strings

grzesiek2010 committed:
Moved the background location option to the popup menu

grzesiek2010 committed:
Final code improvements

grzesiek2010 committed:
Fixed duplicated logs

grzesiek2010 commented:
@yanokwa @zestyping @smoyth 

> I noticed one thing, though. Java type names should be singular; the enum should be EventType rather than EventTypes.

Fixed.

> can you please update the text to what @smoyth suggested at #2772 (comment).

Updated.

> We just want to move it into the pop up menu. @grzesiek2010, have you been able to explore if that is doable?

Done.

I updated the test form so please make sure you have that new version.

smoyte commented:
@grzesiek2010 Would you mind posting some screenshots of the new menu look?

yanokwa commented:
![screen shot 2019-01-17 at 07 40 06](https://user-images.githubusercontent.com/32369/51329845-2ea51f00-1a2b-11e9-8f9d-f5673654e5bf.png)

Great work, @grzesiek2010! I'd like to see the icon on the far left followed by "Location Tracking" (to match the snack bar). I'm suggesting the placement based on what I've seen at https://material.io/design/components/menus.html#anatomy, but I'm interested in other examples others have!

I think this PR is close enough, so I'll mark it as ready for testing while we do the last UI tweaks.

smoyte commented:
Agreed, 'Location Tracking' is better.

I think right-aligned may be the norm for mobile. @yanokwa what you linked is for desktop/web UIs I think. See the screenshot I posted higher up...

yanokwa commented:
I saw the screenshot. Do you have a source so we can evaluate the context better?

smoyte commented:
No, sorry. I forgot where I found that :(

zestyping commented:
> Great work, @grzesiek2010! I'd like to see the icon on the far left followed by "Location Tracking" (to match the snack bar).

@yanokwa I was just about to say "Yeah, that sounds perfect"—and then I went looking for examples and found this in Chrome on Android:
![screenshot_20190117-091036](https://user-images.githubusercontent.com/236086/51336228-dd9c2780-1a38-11e9-903a-dff225d4cf7f.png)

If the checkbox stays on the right, there needs to be a decent amount of padding between the text and checkbox.

It's a shame to have this not match the snackbar, though.  Are we open to putting it on the right in the snackbar?

smoyte commented:
Wait what? There is a checkbox in the snackbar too? That seems odd to me... There shouldn't be UI controls in an ephemeral element like that. Will cause user anxiety—how to I get back to that??

zestyping commented:
Sorry, that was what I thought @yanokwa meant when he said "(to match the snack bar)".  Rereading his comment, I'm seeing now that maybe he meant that the text "Location Tracking" should match the snack bar, rather than that the checkbox should match the snack bar.

smoyte commented:
😅

yanokwa commented:
Y'all are confusing me! So to confirm, the two changes we've agreed to make to the [PR's UI](https://user-images.githubusercontent.com/32369/51329845-2ea51f00-1a2b-11e9-8f9d-f5673654e5bf.png).

1. Change the text from `Background location` to `Track Location`
2. Make sure the checkbox, and only the checkbox aligns right.

grzesiek2010 committed:
Naming improvements

grzesiek2010 commented:
> Change the text from Background location to Track Location

Fixed.

>Make sure the checkbox, and only the checkbox aligns right.

Everything is fine here.

grzesiek2010 committed:
Improved AuditEventTest

grzesiek2010 committed:
Added AuditEventLoggerTest

grzesiek2010 committed:
Added AuditEventSaveTaskTest

mmarciniak90 commented:
@grzesiek2010 I noticed only that menu icon is not visible in snackbar on Android 4.2 and 4.1
![screenshot_2019-01-22-16-12-17](https://user-images.githubusercontent.com/3907554/51544619-9e1e7280-1e60-11e9-8dc2-568925d04a46.png)

Verified on Android: 8.1, 7.0, 6.0, 5.1, 4.4, 4.2, 4.1

Verified cases:
* `Track Location` option in menu is visible only when expected
* Form with all question on one page. Location for `group questions /data` is visible
* Adding new form version with log location coordinates in the background
* Using form which logs location coordinates in the background on old collect version
* Manipulated `location-min-interval` and `location-max-age` values
* Google Play Services are not available (Android 4.4)
* Events:
   * location tracking enabled
   * location tracking disabled
   * location providers enabled
   * location providers disabled
   * location permissions granted
   * location permissions not granted
   * <img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/51538534-d4a0c100-1e51-11e9-996a-1583452d9efc.png">

* Snackbar:
   * ![screenshot_2019-01-22-14-43-44](https://user-images.githubusercontent.com/3907554/51538461-a327f580-1e51-11e9-93a6-70c12fa423ad.png)
   * ![screenshot_20190122-142057](https://user-images.githubusercontent.com/3907554/51538462-a327f580-1e51-11e9-919c-594c8f6f1988.png)
  * ![screenshot_20190122-142107](https://user-images.githubusercontent.com/3907554/51538464-a327f580-1e51-11e9-83d6-5bbf505b94c6.png)

* Dialog:
   * <img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/51541308-ca35f580-1e58-11e9-8399-0e0c967f01ae.png">

mmarciniak90 commented:
@opendatakit-bot unlabel "needs testing"

grzesiek2010 commented:
So it's because some font's don't support that char.
The message looks maybe strange but it's still readable.
@yanokwa should we remove that character or maybe leave it as it is since it's probably only android 4?

yanokwa reviewed (commented):
@grzesiek2010 Please fix the merge conflicts and let's get this merged!

As to the glitch with Android 4, try replacing the `⋮` with `&#8942;` and see if that works better. If not, let's file a quick issue to improve it later.

grzesiek2010 committed:
Merge branch 'master' into gps

grzesiek2010 committed:
Fixed conflict

grzesiek2010 commented:
> As to the glitch with Android 4, try replacing the ⋮ with &#8942; and see if that works better. If not, let's file a quick issue to improve it later.

I tried yesterday and no avail. A new issue is added https://github.com/opendatakit/collect/issues/2813
